### DISCUSSION TITLE
get every item in `mp_obj_list_t`

### DISCUSSION THREAD

**Discussion by tejalbarnwal** [micropython/micropython#9650] [Using MicroPython]
Hey, I am actually trying to store a value from micro python list into a temporary variable. How can I do it?
For example, let say I have a list pointer variable named `input_list` which was declared by `mp_obj_list_t *input_list;`
Now I wish to get the elements 
```cpp
for (size_t i = 0; i < input_list->len; i++) {
       float temp = (float)(input_list->items[i]);
}
```
But when i try to compiled the library, i get the following error:
```
error: pointer value used where a floating-point was expected
                 float temp = (float)(arg->input_list->items[size]);
                 ^~~~~
```

What is the correct way of storing value in `temp` variable, assuming that list is only made of integers or floating values?

**Comment by jimmo** [comment]
In MicroPython, every object is an `mp_obj_t`. This is an opaque type that encodes strings, integers, floats, or pointers-to-objects.

So you need to know what type you expect. In this case, float, so use `mp_float_t temp = mp_obj_float_get(input_list->items[i])`

Note the mp_float_t is important, because it depends on whether MicroPython is built for single or double precision float as to what the built-in type will be.

**Reply by tejalbarnwal** [reply]
Hey, thankyou so much for the quick response!

**Reply by tejalbarnwal** [reply]
Hey @jimmo , I encountered something while i was debugging my code.

The function that I defined in a user defined module is something along the following lines:
```cpp
STATIC mp_obj_t py_tf_regression(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
{
    // did something with args[0] which works fine
   
    // reading the remaining three elements from the args

    // args[1] passed is an integer
    printf("printing arg 1: \t");
    printf("%d \n", (int)mp_obj_float_get(args[1]));

    // args[2] passed is an integer
    printf("printing arg 2: \t");
    printf("%d \n", (int) args[2]);

    //args[3] passed is list of integers
    mp_obj_list_t *arg_list = args[3];
    printf("read into arg list from arg 3 \n");
    for (size_t i = 0; i < arg_list->len; i++) {
        printf( "%d \t", (int) mp_obj_float_get(arg_list->items[i]));
    }
    printf("\n printed arg 3 \n");

}
```

Micropython code for calling the function:
```python
a = 1
b = 1
input_list = [-8, -5, -8,  2]
output = tf.regression(net, a, b, input_list)
```

The output that I get:
```
printing arg 1:         2147483647 
printing arg 2:         3
read into arg list from arg 2 
2147483647      2147483647      2147483647
 printed arg 2
```
Doubts:
1. Why does argument 1 becomes a garbage value? How can I read it correctly?
2. I found that if I initialize `b` as 3, it printed as `5`. I ran the function with other values of `b` as well, noticing the printed value of is always `2*b+1`. What do you think what could be causing it?
3. Further, the list that I passed also contains garbage values. How can I correct this?

**Reply by stinos** [reply]
An integer doesn't have the same representation as a floating point value, so using `mp_float_get` on an `mp_obj_t` which is actually an integer isn't going to work,  you should check what type the argument is using `mp_obj_is_xxx`, then use `mp_obj_get_int` (or similar), see obj.h.

**Reply by v923z** [reply]
You could also use https://github.com/micropython/micropython/blob/b41aaaa8a918a6645ebc6bfa4483bd17286f9263/py/obj.c#L326, which makes sure that you read an integer, if the object is indeed an integer.

**Comment by tejalbarnwal** [comment]
It would be fantastic if  [this](https://micropython-usermod.readthedocs.io/en/latest/) could be added in the documentation under internals [here](https://docs.micropython.org/en/latest/develop/library.html). 
Its a really, really great documentation with a lot of examples.
Thankyou @v923z
